doc/mgmt/SONiC_ConfigDB_support_Yang_default_values_and_profiles.md

```diff
@@ -0,0 +1,392 @@
+
+## 1.2 swss-common return profile from profile DB
+
+- Buffer profile stored in profile tables.
+
+- Return profile is optional.
+  
+  - Application can decide read config with profile or not:
+    - When application read profile:
+      - If user overwrite the profile, user config will be return.
+      - If user not overwrite the profile, profile will be return.
+      - If user 'delete' the profile, will return nothing.
+    - When application not read profile, API will only return user config.
+      - If user overwrite the profile, user config will be return.
+      - If user not overwrite the profile, will return nothing.
+      - If user 'delete' the profile, then there also will no user config exist, will return nothing.
+
+- Backward compatibility with existing code and applications.
+  
+  - For backward compatibility when initialize buffer config from minigraph, config will be write to both config tables and profile tables.
+  - After all code migrate to use profile DB, profile will only write to profile tables.
+  - Profile support delete/revert operation:
+    - In some user scenario, user need delete a profile, and also may add config back later.
+    - For delete operation, data in profile table will not be delete, profile use PROFILE_DELETE table to handle delete/revert:
+      - When delete a profile item, profile provider will add the item key to PROFILE_DELETE table.
+      - When read profile, any key in PROFILE_DELETE will not exist in result.
+      - When user set deleted item back, the item key will be remove from PROFILE_DELETE table.
+      - For example:
+        - BUFFER_POOL table are profile:
+          
+          ```
+          "BUFFER_POOL": {
+          "ingress_lossless_pool": {
+            "size": "26531072",
+            "type": "ingress",
+            "mode": "dynamic",
+            "xoff": "6291456"
+          },
+          "egress_lossless_pool": {
+            "size": "32822528",
+            "type": "egress",
+            "mode": "static"
+          }
+          }
+          ```
+        - On mellanox devices, when set buffer mode to 'dynamic' mode, buffer manager will use swsscommon API to delete buffer pool to reclaim unused buffer resource.
+        - When delete a profile item 'ingress_lossless_pool', the key of this item will be write to PROFILE_DELETE table. and swsscommon read API will not return this item anymore.
+        - Revert operation: when user set buffer mode back, buffer manager will write the deleted item back, when this happen, the key of the item will be removed from PROFILE_DELETE table. and swsscommon read API will return this item.
+
+# 2 Design
+
+- Config DB design diagram:
+
+<img src="./images/swss-common-layer.png"  />
+
+- API design diagram:
+
+<img src="./images/swss-common-default-value.png"  />
+
+### Current design:
+
+- Existing API keeps no change. 
+- Add decorator API to return default value and profile data.
+- Load yang model as lazy as possible.
```

Comment:
Maybe we can cache the parsed yang model information somewhere globally so that we just need to parse the yang model once and can use them many times.
But I'm not sure whether it is worth doing so since caching/storing it also takes time.

Reply:
Fixed by add more detail, current design is to only load once and cache it.